### PR TITLE
Detection of repo conflicts

### LINKED ISSUE #1521: project conflicts don't take into account repo differences
Two projects with identical repo references but different repos will be considered "the same", which is a serious bug.

### PR DESCRIPTION
**Pull request by odisseus**
Fixes #1521 (I hope).

### CODE REVIEW

File: src/core/layer.scala [hunk 1]
```diff
@@ -54,9 +54,11 @@ case class Layer(version: Int,
     })
   })
 
-  def localUniverse(hierarchy: Hierarchy, path: Pointer): Universe = Universe(
+  def localUniverse(hierarchy: Hierarchy, path: Pointer): Try[Universe] = for {
+    foobar <- projects.to[List].traverse { project => project.projectRef(hierarchy) >> (Uniqueness.Unique(_, Set(path))) >> (project.id -> _) }
+  } yield Universe(
     hierarchy = hierarchy,
-    projects = projects.map { project => project.id -> Uniqueness.Unique(project.projectRef, Set(path)) }.toMap,
+    projects = foobar.toMap,
```
Comment (propensive) [inline src/core/layer.scala]: We still need better names than `foobar`...

File: src/core/project.scala [hunk 2]
```diff
@@ -43,15 +43,24 @@ case class Project(id: ProjectId,
...
   }
 
+  private[this] def digesterInput(hierarchy: Hierarchy, m: Module) = {
+    for {
+      universe <- hierarchy.universe
+      rs <- universe.repoSources(m.ref(this))
+    } yield {
+      (m.id, m.kind, m.manifest, m.compiler, m.dependencies.to[List], m.opts.to[List], m.binaries.to[List],
+        m.environment.to[List], m.policy.to[List], m.hidden, m.optDefs.to[List], m.deterministic,
+        (m.sources.to[List] ++ m.resources.to[List]), rs) // FIXME: Resolve these in the hierarchy!
```
Comment (propensive) [inline src/core/project.scala]: I think this is ok. I've read it a few times.

File: src/core/universe.scala [hunk 3]
```diff
@@ -83,16 +83,19 @@ case class Universe(hierarchy: Hierarchy,
     case (acc, (_, Ambiguous(origins))) => acc ++ origins.values
   }
 
-  def checkout(ref: ModuleRef, hierarchy: Hierarchy, layout: Layout)(implicit log: Log): Try[Snapshots] = for {
+  def repoSources(ref: ModuleRef): Try[List[(Repo, List[RepoSource])]] = for {
```
Comment (propensive) [inline src/core/universe.scala]: The return type suggests the order is significant. Shouldn't this be a `Map`?

File: test/pending/commits-conflict/script [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+fury universe projects
+
+fury layer select -l /magnolia
+fury repo list --raw --column Commit
+fury repo update -r magnolia -t v0.17.0
+fury repo list --raw --column Commit
+
+fury layer select -l /
+fury universe imports list #--raw --column Import
+fury universe projects
```
Comment (propensive) [inline test/pending/commits-conflict/script]: Sholudn't this line produce some output? I can't see anything in the `check` file.

**Comment by odisseus** [comment]
I accidentally a circular dependency between `ProjectRef` and `Universe`. Now I have to ditch everything and start over.

Basically, you cannot get the state of the repos while constructing the project, because this requires the layer to be already constructed, including the current project.

This also means that you cannot check whether all the project have consistent requirements to the state of repos in the universe until the universe is completely constructed.

File: test/pending/commits-conflict/script (hunk 4 above)
Comment (odisseus) [reply test/pending/commits-conflict/script]: The test isn't complete yet.

**Comment by propensive** [comment]
I'm not sure it should be an issue. Repos are scoped to the current layer, and we don't need a `Universe` to construct each project, in the context of its layer, but the method which constructs the universe probably needs to construct the `RepoSet`s first, and then pass these as a parameter to the line which constructs the `ProjectEntity`s.

File: src/core/layer.scala (hunk 1 above)
Comment (odisseus) [reply src/core/layer.scala]: Refactored

File: src/core/project.scala (hunk 2 above)
Comment (odisseus) [reply src/core/project.scala]: I just copied this tuple as it was, and added a new element at the end.

File: src/core/universe.scala (hunk 3 above)
Comment (odisseus) [reply src/core/universe.scala]: Refactored

File: test/pending/commits-conflict/script (hunk 4 above)
Comment (odisseus) [reply test/pending/commits-conflict/script]: Fixed

File: test/passing/projects-proliferate/script [hunk 5]
```diff
@@ -41,6 +41,6 @@ fury layer import -l $IPFSREF2 --name platform2
 echo $?
 fury layer list --raw --column ref
 fury universe projects list --raw
-fury universe projects proliferate -p hello-world@8803c7
+fury universe projects proliferate -p hello-world@9a328e
 fury layer list --raw --column ref
 fury universe projects list --raw
```
Comment (propensive) [inline test/passing/projects-proliferate/script:46]: Shouldn't this line list the `hello-world` project, without the `@hash`?

**Review by propensive — APPROVED** [review]

File: test/passing/projects-proliferate/script (hunk 5 above)
Comment (odisseus) [reply test/passing/projects-proliferate/script:46]: It does. The command at line 46 produces the last three lines in the check file:
```
U2Asaiwr
hello-world
scala
```

**Pull request merged**